plug-ins/gpc/gpc.3
.TH gpc 3 "30 Apr 1998" "Roadkills-R-Us" "Gimp Plug-ins"
.\"Copyright 1997-8, Roadkills-R-Us, Austin, TX, USA. All rights reserved.
.\"
.\" gpc version 1.4
.de Ss
.sp
.ft CW
.nf
..
.de Se
.fi
.ft P
.sp
..

.SH NAME
GPC \- GTK Plug-in Convenience library

.SH SYNOPSIS
.Ss
#include <plug-ins/gpc/gpc.h>
.Se
.Ss
void gpc_setup_tooltips(GtkWidget *parent);
.Se
Initialize tooltips interface, set colors.
.Ss
void gpc_set_tooltip(GtkWidget *widget, const char *tip);
.Se
Set tooltip for a widget (if tip is non-null).

.Ss
void gpc_add_action_button(char *label,
    GtkSignalFunc callback, GtkWidget *dialog, char *tip);
.Se
Add action button (with tooltip) to a dialog.
.Ss
void gpc_add_radio_button(GSList **group, char *label,
    GtkWidget *box, gint *value, char *tip);
.Se
Add radio button (with tooltip) to a dialog.
.Ss
void gpc_add_hscale(GtkWidget *table, int width,
    float low, float high, gdouble *val,
    int left, int right, int top, int bottom, char *tip);
.Se
Add horizontal scale widget (with tooltip) to a dialog at given location.
.Ss
void gpc_add_label(char *value, GtkWidget *parent,
    int left, int right, int top, int bottom);
.Se
Add label widget (no tooltip) to a dialog at given location.

.Ss
void gpc_close_callback(GtkWidget *widget, gpointer data);
.Se
Destroy callback - quit this plug-in.
.Ss
void gpc_cancel_callback(GtkWidget *widget, gpointer data);
.Se
Cancel button callback - go away without saving state, etc.
.Ss
void gpc_scale_update(GtkAdjustment *adjustment, double *scale_val);
.Se
Scale update callback - update the SCALE widget's data.
.Ss
void gpc_text_update(GtkWidget *widget, gpointer data);
.Se
Text update callback - update the TEXT widget's data.

.SH DESCRIPTION
This is a set of routines to make life easier on plug-in writers.
It helps keep the GUI code short and sweet. (In the plug-in for
which it was originally designed, it cut the GUI code in half.)
It's somewhat arbitrary in what it includes so far, because I
haven't needed everything in GTK.  Feel free to contribute more
to it.

.SS TOOLTIPS ROUTINES

.B gpc_setup_tooltips()
must be called first.  This initializes internal data
structures and sets the tooltip colors.  It can be called
with any widget high enough in the hierarchy to contain
all the widgets needing tooltips.  Typically this will be
the container widget under the top-level frame:

.nf
Example:
    frame = gtk_frame_new("Parameter Settings");
        :
    gtk_box_pack_start(GTK_BOX(GTK_DIALOG(dlg)->vbox),
        frame, TRUE, TRUE, 0);
    table = gtk_table_new(4, 2, FALSE);
        :
    gtk_widget_show(table);
    gpc_setup_tooltips(table);
.fi

.PP
.B gpc_set_tooltip()
may be called directly, but is usually called inside other convenience
functions.  If called directly, it must still be after the call to
gpc_setup_tooltips().  It hides a lot of detail of the GTK tooltips:

.nf
Example:
    gtk_widget_show(button);
    gpc_set_tooltip(button, tip);
.fi



.SS USER INTERFACE ROUTINES

These routines all hide implementation details to make it
easier to lay out a GUI with a consitent, gimp-style interface,
while keeping the plug-in code cleaner.

.PP
.B gpc_add_action_button()
adds an action button (such as [Cancel]
or [OK] to the action button area of a frame.  The
.I callback
argument has the standard GTK callback interface.  A standard
callback is provided for [Cancel] buttons if you wish to use it.

.PP
Usage:
.nf
    void
    gpc_add_action_button(
        char *label,            /* text for action button */
        GtkSignalFunc callback, /* callback function address */
        GtkWidget *dialog,      /* dialog widget to contain button */
        char *tip               /* tooltip text */
    )

Example:
    static void
    randomize_ok_callback(GtkWidget *widget, gpointer data) {
        rndm_int.run = TRUE;
        gtk_widget_destroy(GTK_WIDGET(data));
    }
        :
    gpc_add_action_button("OK",
        (GtkSignalFunc) randomize_ok_callback, dlg,
        "Accept settings and apply filter to image");
    gpc_add_action_button("Cancel",
        (GtkSignalFunc) gpc_cancel_callback, dlg,
        "Close plug-in without making any changes");
.fi

.PP
.B gpc_add_radio_button()
adds a radio button.  If the radio group does not exist,
it will be created and passed back in the
.I group
argument.  A standard callback will be attached to the radio
button, requiring a state variable which you provide via the
.I value
argument.

.nf
Usage:
    void
    gpc_add_radio_button(
        GSList **group,         /* address of radio group */
        char *label,            /* label for new radio button */
        GtkWidget *box,         /* radio box for this radio button */
        gint *value,            /* address of state variable */
        char *tip               /* tooltip text */
    )

Example:
    GSList *type_group = NULL;
        :
    gpc_add_label("Randomization Type:", table, 0, 1, 0, 1);

    toggle_hbox = gtk_hbox_new(FALSE, 5);
    gtk_container_border_width(GTK_CONTAINER(toggle_hbox), 5);
    gtk_table_attach(GTK_TABLE(table), toggle_hbox, 1, 2, 0, 1,
        GTK_FILL | GTK_EXPAND, GTK_FILL, 5, 0);

    gpc_add_radio_button(&type_group,
        "Hurl", toggle_hbox, &do_hurl,
        "Hurl random colors onto pixels");
    gpc_add_radio_button(&type_group,
        "Pick", toggle_hbox, &do_pick,
        "Pick at random from neighboring pixels");
    gpc_add_radio_button(&type_group,
        "Slur", toggle_hbox, &do_slur,
        "Simplistic melt");
.fi


.PP
.B gpc_add_hscale()
adds a horizontal scale to a table container at the designated coordinates.
A standard callback will be attached to the scale,
requiring a state variable which you provide via the
.I value
argument.

.nf
Usage:
    void
    gpc_add_hscale(
        GtkWidget *table,       /* table widget to hold scale */
        int width,              /* width (in pixels) of scale */
        float low,              /* low value for scale */
        float high,             /* high value for scale */
        gdouble *value,         /* pointer to current value */
        int left,               /* left table position info */
        int right,              /* right table position info */
        int top,                /* top table position info */
        int bottom,             /* bottom table position info */
        char *tip               /* tooltip text */
    )

Example:
    gpc_add_label("Randomization %:", table, 0, 1, 2, 3);
    gpc_add_hscale(table, SCALE_WIDTH,
        1.0, 100.0, &pivals.rndm_pct, 1, 2, 2, 3,
        "Percentage of pixels to be filtered");
.fi

.PP
.B gpc_add_label()
simply adds a label at the designated coordinates in the table.
Labels don't get tooltips.

.nf
Usage:
    void
    gpc_add_label(
        char *value,            /* text for new label */
        GtkWidget *table,       /* table widget to hold label */
        int left,               /* left table position info */
        int right,              /* right table position info */
        int top,                /* top table position info */
        int bottom              /* bottom table position info */
    )

Example:
    gpc_add_label("Randomization %:", table, 0, 1, 2, 3);
.fi

.SS CALLBACKS:

.B gpc_close_callback()
is used in OK callbacks, and anywhere else
you need a callback to destroy widgets.
The default cancel callback,
.B gpc_cancel_callback()
simply closes (destroys) the current panel.
The
.B gpc_scale_update()
and
.B gpc_text_update()
callbacks update the appropriate widget's data from
that widget.

.DIAGNOSTICS

No special diagnostics are provided.

.SH BUGS

This software should be 100% Bug-Free [tm].

.SH AUTHOR

Miles O'Neal
.br
<meo@rru.com>
.br
http://www.rru.com/~meo/
.br
Leander, TX
.br
Additionally, some of the code may have been distilled from
the following plug-ins:
.I alienmap
(Copyright (C) 1996, 1997 Daniel Cotting)
.I plasma
(Copyright (C) 1996 Stephen Norris),
.I oilify
(Copyright (C) 1996 Torsten Martinsen),
.I ripple
(Copyright (C) 1997 Brian Degenhardt) and
.I whirl
(Copyright (C) 1997 Federico Mena Quintero).


NEWS
                   ------------------------------
                   GNU Image Manipulation Program
                         Development Branch
                   ------------------------------

This is the development branch of GIMP.

Overview of Changes from GIMP 3.2.4 to GIMP 3.3.2
=================================================

Core:

  * Using native file choosers in various parts of the code.
  * Rendering MyPaint brush previews at best sizes to be non-blurry.
  * GimpViewable preview API now passes around the scale factor to allow
    best decision when rendering previews.
  * Error console save dialog now has a ".log" extension filter.
  * Improve list of "Recent Images" in the Welcome dialog, with no
    arbirtrary limitation of 8 rows.
  * More GtkTreeView replaced in preparation of GTK4 port.
  * Various preliminatory changes in preparation of GTK4 port.
  * macOS: touchpads are now less sensitive in pan gestures and behave as
    expected.
  * MSIX: it is now possible to open multiple images from the OS.
  * Stop noninteractive filters on invisible layers unless the relevant
    option is enabled in Preferences.
  * Various layer modes are now being ported to work on non-RGBA
    contents when it makes sense.
  * Non-GUI filters will be applied non-destructively on group layers as
    well as non-raster layers. For now they are still applied
    destructively only on raster layers, until a proper design decision
    is taken on this topic.
  * "Layers to Image Size" action is now only sensitive on raster
    layers.
  * Layer Masks can now have non-destructive filters (listed separately
    in the NDE popover).
  * Added "Rotation" input for dynamics. This gives us support of the
    Wacom Art Pen 2 stylus at least.
  * We now verify if a plug-in has been updated since last start by
    using the changed time (ctime) as fallback when the modified time
    (mtime) is 0. This is because on some environment (such as ostree,
    hence flatpak), all files mtimes are set to the UNIX epoch.
  * Added support of dashboard backtraces on Windows ARM64.
  * Krita KPL palette import support added.
  * Work started on a new project format.
  * Do not perform the (wrong) optimization replacing pass-through by
    the Normal layer mode for layers with effects.
  * Dashboard backtraces now support macOS.
  * Clear does not affect non-raster layers, unless rasterized.
  * New project format for GIMP (meant to replace XCF) is under
    development. Sub-Features:
    - More robust image recovery (e.g. after a crash, external energy
      loss or other issues) reimplemented based on the new format.
    - Constant, reliable and seamless auto-save of opened images.
  * About dialog: more info about the OS, dependencies and hardware is
    being shown. A "Copy" button allows to easily add the
    markdown-formatted info to clipboard for pasting in reports.
  * Improve internal logics to determine if an item is "writable"
    regarding actions modifying item contents.
  * "Reshow Filter" and "Reapply Filter" now work on group layers too.

Tools:

  * Gradient tool: new option "Editable gradient" to apply the gradient
    non-destructively.
  * Heal tool: healing behaves more as expected near edges.
  * Text tool:
    - Allow selecting with Shift-Click.
    - Preview text layer content while Alt-dragging.
    - Ignore selected layer mask of a text layer when selecting it with
      the tool to edit the text contents. The layer mask still stays the
      selected item in such a case.
  * Transform tools: add a tooltip on the chain icon to better explain
    what it does.
  * MyPaint Brush tool: new "Spectral Blending" checkbox and "Pigment"
    slider to blend spectrally while painting.
  * Color Picker tool: there was an exception ignoring Sample Merged
    option when a single layer was visible (allowing to retrieve the
    exact palette color from indexed images). This exception has been
    further restricted to the absence of filters. Sample Merged, even of
    single layer images, should return the color after filters are
    applied.
  * Path tool: new "Auto-connect on click" checkbox setting (ON by
    default) allowing to decide whether to auto-connect, or not, when
    clicking on the end node (instead allowing users to readjust it).

Graphical User Interface:

  * Small reorganization in GimpStrokeEditor.
  * Welcome dialog: icons and links realigned on Welcome tab; widgets
    realigned in Contribute tab.
  * New "branding" colors added in AppStream metadata.
  * macOS: title bar follows scheme. This is true for native file
    dialogs on macOS as well.
  * Levels tool now listens to the icon size setting.
  * Roadmap menu item removed from Help > GIMP Online menu.
  * We do not rotate the checkerboard anymore when rotating the canvas,
    but also scrolling or zooming. This significantly improves performance
    when interacting with a rotated canvas.
  * GimpColorScale moved from hardcoded (bad!) to proper theme-based CSS
    styling.
  * Append "(read only)" to system resource (brushes, gradients,
    palettes…) titles when trying to edit them in their respective
    Editor dockables.
  * Action Search dialog:
    - The shortcuts hints are moved to the right side, similar to the
      rest of the application (e.g. shortcuts hints in menus).
    - A ".search-results" class was added for the results row, and theme
      was updated to highlight the hovered action row.
  * Icon picker widget: the file filter name is now titled "Image
    files".
  * Multiple selected layers are now added to a new group.
  * Navigation dockable: grab back focus on the view, after using the
    zoom slider.
  * Patterns which are smaller than the preview area (in 1 or 2
    dimensions) will now fill the full area by repeating themselves,
    since they are meant to be tileable anyway. It makes for nicer
    pattern previews.
  * Size entry in Scale and Print Size dialogs receive the focus by

plug-ins/randomize/gpc.c
/****************************************************************************
 * This is a convenience library for plugins for the GIMP v 0.99.8 or later.
 * Documentation is available at http://www.rru.com/~meo/gimp/ .
 *
 * Copyright (C) 1997 Miles O'Neal  <meo@rru.com>  http://www.rru.com/~meo/
 * Blur code Copyright (C) 1995 Spencer Kimball and Peter Mattis
 * GUI based on GTK code from:
 *    alienmap (Copyright (C) 1996, 1997 Daniel Cotting)
 *    plasma   (Copyright (C) 1996 Stephen Norris),
 *    oilify   (Copyright (C) 1996 Torsten Martinsen),
 *    ripple   (Copyright (C) 1997 Brian Degenhardt) and
 *    whirl    (Copyright (C) 1997 Federico Mena Quintero).
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 ****************************************************************************/

/****************************************************************************
 * gpc:
 *
 * gpc version 1.1 (3 Feb 1998, MEO)
 * history
 *     1.1 -  3 Feb 1998 MEO
 *         removed tooltips from action buttons
 *     1.0 -  2 Feb 1998 MEO
 *         FCS
 *
 * Please send any patches or suggestions to the author: meo@rru.com .
 * 
 ****************************************************************************/

#include <stdlib.h>
#include <time.h>
#include "libgimp/gimp.h"
#include "gtk/gtk.h"


/*
 *  TOGGLE UPDATE callback
 */
static void
gpc_toggle_update(GtkWidget *widget, gpointer data) {
    int *toggle_val;

    toggle_val = (int *) data;

    if (GTK_TOGGLE_BUTTON(widget)->active)
      *toggle_val = TRUE;
    else
      *toggle_val = FALSE;
}
/*
 *  DESTROY callback
 */
void
gpc_close_callback(GtkWidget *widget, gpointer data) {
    gtk_main_quit();
}

/*
 *  CANCEL BUTTON callback
 */
void
gpc_cancel_callback(GtkWidget *widget, gpointer data) {
    gtk_widget_destroy(GTK_WIDGET(data));
}

/*
 *  SCALE UPDATE callback
 */
void
gpc_scale_update(GtkAdjustment *adjustment, double *scale_val) {
    *scale_val = adjustment->value;
}


/*
 *  TEXT UPDATE callback
 */
void
gpc_text_update(GtkWidget *widget, gpointer data) {
  gint *text_val;

  text_val = (gint *) data;

  *text_val = atoi(gtk_entry_get_text(GTK_ENTRY(widget)));
}



/*
 *  TOOLTIPS ROUTINES
 */
static GtkTooltips *tips;


void
gpc_setup_tooltips(GtkWidget *parent)
{
    static GdkColor tips_fg, tips_bg;

    tips = gtk_tooltips_new();

    /* black as foreground: */

    tips_fg.red   = 0;
    tips_fg.green = 0;
    tips_fg.blue  = 0;
    gdk_color_alloc(gtk_widget_get_colormap(parent), &tips_fg);

    /* postit yellow (khaki) as background: */

    tips_bg.red   = 61669;
    tips_bg.green = 59113;
    tips_bg.blue  = 35979;
    gdk_color_alloc(gtk_widget_get_colormap(parent), &tips_bg);

    gtk_tooltips_set_colors(tips, &tips_bg, &tips_fg);
}


void
gpc_set_tooltip(GtkWidget *widget, const char *tip)
{
    if (tip && tip[0])
        gtk_tooltips_set_tips(tips, widget, (char *) tip);
}

void
gpc_add_action_button(char *label, GtkSignalFunc callback, GtkWidget *dialog
    /* , char *tip */ )
{
    GtkWidget *button;

    button = gtk_button_new_with_label(label);
    GTK_WIDGET_SET_FLAGS(button, GTK_CAN_DEFAULT);
    gtk_signal_connect(GTK_OBJECT(button), "clicked", callback, dialog);
    gtk_box_pack_start(GTK_BOX(GTK_DIALOG(dialog)->action_area),
	button, TRUE, TRUE, 0);
    gtk_widget_grab_default(button);
    gtk_widget_show(button);
    /* gpc_set_tooltip(button, tip); */
}

void
gpc_add_radio_button(GSList **group, char *label, GtkWidget *box,
    gint *value, char *tip)
{ 
    GtkWidget *toggle;

    toggle = gtk_radio_button_new_with_label(*group, label);
    *group = gtk_radio_button_group(GTK_RADIO_BUTTON(toggle));
    gtk_box_pack_start(GTK_BOX(box), toggle, FALSE, FALSE, 0);
    gtk_signal_connect(GTK_OBJECT(toggle), "toggled",
        (GtkSignalFunc) gpc_toggle_update, value);
    gtk_toggle_button_set_state(GTK_TOGGLE_BUTTON(toggle), *value);
    gtk_widget_show(toggle);
    gtk_widget_show(box);
    gpc_set_tooltip(toggle, tip);
}

void
gpc_add_label(char *value, GtkWidget *table, int left, int right,
    int top, int bottom)
{
    GtkWidget *label;

    label = gtk_label_new(value);
    gtk_misc_set_alignment(GTK_MISC(label), 0.0, 0.5);
    gtk_table_attach(GTK_TABLE(table), label, left, right, top, bottom,
        GTK_FILL | GTK_EXPAND, GTK_FILL, 5, 0);
    gtk_widget_show(label);
}

void
gpc_add_hscale(GtkWidget *table, int width, float low, float high,
    gdouble *val, int left, int right, int top, int bottom, char *tip)
{
    GtkWidget *scale;
    GtkObject *scale_data;

    scale_data = gtk_adjustment_new(*val, low, high, 1.0, 1.0, 0.0);
    scale = gtk_hscale_new(GTK_ADJUSTMENT(scale_data));
    gtk_widget_set_usize(scale, width, 0);
    gtk_table_attach(GTK_TABLE(table), scale, left, right, top, bottom,
        GTK_FILL | GTK_EXPAND, GTK_FILL, 0, 0);
    gtk_scale_set_value_pos(GTK_SCALE(scale), GTK_POS_TOP);
    gtk_scale_set_digits(GTK_SCALE(scale), 0);
    gtk_range_set_update_policy(GTK_RANGE(scale), GTK_UPDATE_DELAYED);
    gtk_signal_connect(GTK_OBJECT(scale_data), "value_changed",
        (GtkSignalFunc) gpc_scale_update, val);
    gtk_widget_show(scale);
    gpc_set_tooltip(scale, tip);
}
